README.md
## dozer111/PhpMoreFunctions

### About
Library, which add some new global php functions, or add new functions, 
which has they`re prototype in pure php  

**[Full list of funtions](https://github.com/dozer111/additional_functions/wiki/Functions-list)**


### Install/Usage
```bash
composer require dozer111/php_more_functions
```
```php
// require vendor/autoload, if you had no autoload on file

$someTestData = '0';

....
....
....

if(emptiest($someTestData))
{
    ....
    ....
}
```

### Help me
So, may you have some good new ideas/functuions, or can do codeReview
to do it better, please, write me on `alkhonko@gmail.com`

CHANGELOG.md
## 2.0.0
#### Changed function names
* `floatval_normalized` => `normalizeFloatval`
* `implode_and_normalize` => `normalizeImplode`


## 2.1.0
* added [truncateNumber()](https://github.com/dozer111/additional_functions/blob/master/src/textProcessing/strings.php#L93) function


## 2.2 
* added [toArray()](https://github.com/dozer111/additional_functions/blob/master/src/variable_and_type_related_extensions/variable_handling.php#L73)


## 2.2.1 
* fix truncateNumber() function, which return wrong value when number had e
```php
// was
$number = 123.12313123131e5;
$res = truncateNumber($number); //123
// fixed
$res = truncateNumber($number); //123e5
```


























tests/unit/ToArrayTest.php
<?php

use Codeception\Specify;

class ToArrayTest extends \Codeception\Test\Unit
{
    /**
     * @var \UnitTester
     */
    protected $tester;

    use Specify;



    protected function _before()
    {
    }

    protected function _after()
    {
    }

    public function testValueIsArray(){
        $data = [
            [],
            array(),
            [1,2,3,4]
        ];


        foreach ($data as $testCase){
            $toArrayVal = toArray($testCase);
            $this->tester->assertIsArray($toArrayVal);
            $this->tester->assertSame($testCase,$toArrayVal);
        }


    }

    public function testValueIsNotArray(){
        $data = [
            1,
            'asd',
            null,
            false,
            new DateTime(),
        ];

        foreach ($data as $testCase){
            $toArrayVal = toArray($testCase);
            $this->tester->assertIsArray($toArrayVal);
        }

    }
}

tests/unit/EmptiestTest.php
<?php

use Codeception\Test\Unit;

/**
 * This test suite shall coverage https://www.php.net/manual/ru/function.empty.php
 *
 * The main idea is, that we want to get from emptiest() function real bool value, whether it`s empty
 */

class EmptiestTest extends Unit
{
    /**
     * @var \UnitTester
     */
    protected $tester;

    private $resource;

    protected function _before()
    {
        $filePath = __DIR__ . '/../_data/text.txt';

        $this->resource = fopen($filePath, 'rb');
    }

    protected function _after()
    {
        fclose($this->resource);
    }

    public function testEmptyString()
    {
        $emptyString = '';

        $this->assertTrue(emptiest($emptyString));
    }

    public function testNotEmptyString()
    {
        $notEmptyString1 = ' ';
        $notEmptyString2 = 'some test text';

        $this->assertFalse(emptiest($notEmptyString1));
        $this->assertFalse(emptiest($notEmptyString2));
    }

    public function testOnPasteZero()
    {
        $input1 = 0;
        $input2 = 0.0;
        $input3 = '0';

        $this->assertFalse(emptiest($input1));
        $this->assertFalse(emptiest($input2));
        $this->assertFalse(emptiest($input3));
    }

    public function testOnPasteNull()
    {
        $input = null;
        $this->assertTrue(emptiest($input));
    }

    public function testOnPasteBool()
    {
        $input = false;
        $input2 = true;
        $this->assertFalse(emptiest($input));
        $this->assertFalse(emptiest($input2));
    }

    public function testOnPasteArray()
    {
        $emptyArray = [];
        $notEmptyArray = [1, 2, 3, 4, 5];


        $this->assertTrue(emptiest($emptyArray));
        $this->assertFalse(emptiest($notEmptyArray));


    }

    public function testOnPasteResource()
    {
        $resource = $this->resource;

        $this->assertFalse(emptiest($resource));

    }

    public function testOnPasteObject()
    {
        $object = new ArrayObject();

        $this->assertFalse(emptiest($object));
    }


}

tests/unit/TruncateNumberTest.php
<?php 
class TruncateNumberTest extends \Codeception\Test\Unit
{
    /**
     * @var \UnitTester
     */
    protected $tester;

    use \Codeception\Specify;


    /**
     * @dataProvider dataValid
     * @param $input
     * @param $expected
     * @param $decimals
     * @return void
     */
    public function testDataValid($input,$expected,$decimals)
    {
        expect(truncateNumber($input,$decimals))->same($expected);
    }

    public function dataValid()
    {
        $testInt = 123456;
        $testFloat = 123456.129951474859;

        $testFloatWithE = '123456.129951474859e7';
        $testFloatWithEMinus = '123456.129951474859e-7';

        $testFloatWithBigE = '123456.129951474859E7';
        $testFloatWithBigEMinus = '123456.129951474859E-7';

        $testFloatWithEAsFloat = 123456.129951474859e3;
        $tooBigDecAsString = '123456.1234567890123456789012345678901';
        $tooBigDecAsStringWithE = '123456.1234567890123456789012345678901e4';
        // input,expected,decimals
        return [
            'test on int' => [$testInt,strval($testInt),0],

            'test on float 1' => [$testFloat,'123456.12',2],
            'test on float 2' =>[$testFloat,'123456.1299',4],
            'test on float 3' =>[$testFloat,'123456.12995',5],

            'test on float with e 1' =>[$testFloatWithE,'123456.12e7',2],
            'test on float with e 2' =>[$testFloatWithE,'123456.1299e7',4],
            'test on float with e 3' =>[$testFloatWithE,'123456.12995e7',5],
            'test on float with e 11' =>[$testFloatWithEMinus,'123456.12e-7',2],
            'test on float with e 12' =>[$testFloatWithEMinus,'123456.1299e-7',4],
            'test on float with e 13' =>[$testFloatWithEMinus,'123456.12995e-7',5],


            'test on float with big e 1' =>[$testFloatWithBigE,'123456.12E7',2],
            'test on float with big e 2' =>[$testFloatWithBigE,'123456.1299E7',4],
            'test on float with big e 3' =>[$testFloatWithBigE,'123456.12995E7',5],
            'test on float with big e 11' =>[$testFloatWithBigEMinus,'123456.12E-7',2],
            'test on float with big e 12' =>[$testFloatWithBigEMinus,'123456.1299E-7',4],
            'test on float with big e 13' =>[$testFloatWithBigEMinus,'123456.12995E-7',5],




            'test on float with sting in end 1' => ['12.75 HP left','12',0],
            'test on float with sting in end 2' => ['12.75 HP left','12.75',2],

            'test on too big decimal value 1' => ['123456.129951474859','123456.129951474859',1e15],
            'test on too big decimal value 2' => [$tooBigDecAsString,'123456.123456789',9],
            'test on too big decimal value 3' => [$tooBigDecAsString,'123456.123456789012',12],
            'test on too big decimal value 4' => [$tooBigDecAsString,'123456.12345678901234567890',20],
            'test on too big decimal value 5' => [123456.129951474859,'123456.12995147',1e15],

            'test on too big decimal value with e 1' => ['123456.129951474859e3','123456.129951474859e3',1e15],
            'test on too big decimal value with e 2' => [$tooBigDecAsStringWithE,'123456.123456789e4',9],
            'test on too big decimal value with e 3' => [$tooBigDecAsStringWithE,'123456.123456789012e4',12],
            'test on too big decimal value with e 4' => [$tooBigDecAsStringWithE,'123456.12345678901234567890e4',20],
            'test on too big decimal value with e 5' => ['123456.129951474859e5','123456.129951474859e5',1e15],

            'test on too big decimal value with e as float 1' => [$testFloatWithEAsFloat,'123456129.95',2],
            'test on too big decimal value with e as float 2' => [$testFloatWithEAsFloat,'123456129.9514',4],
        ];
    }





    public function testDataInvalid()
    {
        $this->specify('$number is pure string',function(){
            $number = 'LOL KEK';
            $this->expectException(InvalidArgumentException::class);
            truncateNumber($number);
        });

        $this->specify('$number is string on start',function(){
            $number = 'LOL KEK 12345';
            $this->expectException(InvalidArgumentException::class);
            truncateNumber($number);
        });
    }

    public function testDataEmpty()
    {
        $emptyData = [
            '',
            null
        ];

        foreach ($emptyData as $testCaseData)
        {
            $this->assertSame(null,truncateNumber($testCaseData));
        }
    }

}

src/variables.php
<?php

/**
 * List of custom functions for variables
 * This is the addition for https://www.php.net/manual/ru/function.empty.php
 */

if (!function_exists('emptiest')) {
    /**
     * Show is the value is really empty
     * Btw we assume, that $value is exists
     *
     * We says, that value is really empty, when
     * 1) null
     * 2) empty array
     * 3) empty string
     *
     * @param $value
     * @return bool
     */
    function emptiest($value): bool
    {
        $type = gettype($value);

        switch ($type) {
            case 'boolean':
            case 'integer':
            case 'double':
            case 'float':
            case 'object':
            case 'resource':
                return false;
            case 'string':
                return !boolval(mb_strlen($value)); // 0 => true, 1+ => false
            case 'array':
                return empty($value);
            default:
                return true;
        }
    }
}

if(!function_exists('normalized_floatval'))
{
    /**
     * Sometimes, we have boolean value in wrong format:
     * 123,123213 instead 123.123213
     *
     * Because of it we have a couple of problems
     * floatval(123,123213) => 123.0
     * yii\i18n\Formatter->normalizeNumericValue('1,5') => InvalidArgumentException
     *
     * This function copy logic from floatval(),and add ',' as delimiter
     */
    function normalized_floatval($value):float
    {
        $value = trim($value);

        if (preg_match('/^\d+.*/', $value)) {

            preg_match('/^\d+[,.]?\d*/',$value,$matches);
            $value = str_replace(',','.',$matches[0]);

        }


        return floatval($value);
    }
}















src/textProcessing/strings.php
<?php

/**
 * List of custom functions for variable handling functions
 * @see https://www.php.net/manual/en/ref.strings.php
 *
 */


if (!function_exists('normalizeImplode')) {


    /**
     * Addition for https://www.php.net/manual/en/function.implode.php
     *
     * 1)Join elements in string
     * 2)remove multiple glue characters
     *
     * Need for situation, when
     * We have some value after implode[1], or concatenation[2]
     * and there is need to check, that imploded symbol is not duplicated
     * /test1/test2/test3/test4/test5 instead of /test1//test2/test3/test4//test5
     *
     *
     * @param string $glue
     * @param array $pieces
     * @return string
     * @example Same idea from yii2
     * @see https://www.yiiframework.com/doc/api/2.0/yii-web-urlnormalizer#$collapseSlashes-detail
     *
     *
     * @example
     * $rootPath = dirname(dirname(__DIR__));
     * $dirPath = SomeObj->getPath();
     * $file = $user->getImage();
     * [1] => $fullPath = implode('/',[$rootPath,$dirPath,$file]);
     * [2] => $fullPath = $rootPath.'/'.$dirPath.'/'.$file;
     *
     * #To prevent this we also do
     * $fullPath = preg_replace('/\s{2,}/',' ',$fullPath);
     *
     * #New solution =>
     * $fullPath = normalizeImplode('/',[$rootPath,$dirPath,$file]);
     *
     *
     *
     */
    function normalizeImplode(string $glue, array $pieces): string
    {

        $string = implode($glue, $pieces);


        $removeDuplicatesClosure = function (string $implodedString, string $duplicateSymbol) {
            $symbolToReplace = $duplicateSymbol;

            /**
             * @see https://www.regular-expressions.info/characters.html#special
             */
            $specialRegexSymbols = [
                '[',
                ']',
                '\\',
                '/',
                '^',
                '$',
                '.',
                '|',
                '?',
                '*',
                '+',
                '(',
                ')',
                '{',
                '}',
            ];

            if (in_array($duplicateSymbol, $specialRegexSymbols)) {
                $duplicateSymbol = "\\$duplicateSymbol";
            }
            return preg_replace("/{$duplicateSymbol}{2,}/", $symbolToReplace, $implodedString);

        };


        return $removeDuplicatesClosure($string, $glue);

    }


}

if (!function_exists('truncateNumber')) {

    /**
     * Same logic as
     * @see https://www.php.net/manual/en/function.number-format
     *
     * Difference is this function don`t round value
     * and give you string value of it
     * or null, if you paste real empty value
     *
     *
     *
     * @important
     * There can be troubles,when u try to paste float val with tooMany digits after comma
     * 123.828282828282828282828
     * It would be better to work with same values as with sting
     *
     *
     * @example
     * $number = 123.459278;
     * $data = truncateNumber($number,2); // 123.45
     * $data = truncateNumber($number,4); // 123.4592
     *
     * @example 2
     * $number = '123456.1234567890123456789012345678901';
     * $value = truncateNumber($number,9); // 123456.123456789
     *
     * @param $number
     * @param int $decimals
     * @return string|null
     */
    function truncateNumber($number, int $decimals = 0):?string
    {

        if(emptiest($number))
            return null;

        if (preg_match('/^-?\d+.*/', $number)) {

            preg_match('/^-?\d+[,.]?\d*/', $number, $matches);
            $numberChanged = str_replace(',', '.', $matches[0]);

        }else
        {
            throw new InvalidArgumentException('$number must be numeric!');
        }


        $decimals = abs($decimals);

        /**
         * @see https://www.php.net/manual/en/regexp.reference.repetition.php
         */
        $pregMatchLimit = 65535;
        $decimals = ($decimals > $pregMatchLimit) ? $pregMatchLimit : $decimals;
        /**
         * @see https://regex101.com/r/xcMNvx/1
         */
        preg_match('/^-?\d+[,.]?\d*(?<e>e-?\d+)?/i',$number,$matches);
        $e = $matches['e'] ?? '';

        $pattern = ($decimals > 0) ? "/^-?\d+(\.\d{1,$decimals})?/" : "/^-?\d+/";
        preg_match($pattern, $numberChanged, $matches);

        $numberWithE = $matches[0].$e;


        return $numberWithE;
    }

}






src/variable_and_type_related_extensions/variable_handling.php
<?php

/**
 * List of custom functions for variable handling functions
 * @see https://www.php.net/manual/en/book.var.php
 *
 */

if (!function_exists('emptiest')) {
    /**
     * Addition for https://www.php.net/manual/ru/function.empty.php
     *
     * Show is the value is really empty
     * Btw we assume, that $value is exists
     *
     * We says, that value is really empty, when
     * 1) null
     * 2) empty array
     * 3) empty string
     *
     * @covers EmptiestTest
     * @param $value
     * @return bool
     */
    function emptiest($value): bool
    {
        $type = gettype($value);

        switch ($type) {
            case 'boolean':
            case 'integer':
            case 'double':
            case 'float':
            case 'object':
            case 'resource':
                return false;
            case 'string':
                return !boolval(mb_strlen($value)); // 0 => true, 1+ => false
            case 'array':
                return empty($value);
            default:
                return true;
        }
    }
}

if(!function_exists('normalizeFloatval'))
{
    /**
     * Sometimes, we have boolean value in wrong format:
     * 123,123213 instead 123.123213
     *
     * Because of it we have a couple of problems
     * floatval(123,123213) => 123.0
     * yii\i18n\Formatter->normalizeNumericValue('1,5') => InvalidArgumentException
     *
     * This function copy logic from floatval(),and add ',' as delimiter
     *
     * @covers NormalizedFloatvalTest
     * @param $value
     * @return float
     */
    function normalizeFloatval($value):float
    {
        $value = truncateNumber($value,1e9);
        return floatval($value);
    }
}




if(!function_exists('toArray'))
{

    /**
     * Give the $value,
     * and if it`s not an array
     *      => wrap to array
     *
     * Difference of this function, and (array)$smth,
     * is that i just wrap, and do nothing specific with value
     * @see https://gist.github.com/dozer111/916283f6ff7f93b63532bd91cc026002
     *
     *
     * @example
     * $from = $data->from; // string|array
     * $from2 = toArray($data->from); // 100% array
     *
     * @covers ToArrayTest
     * @param $value
     * @return array
     */
    function toArray($value):array
    {
        return (is_array($value))
            ? $value
            : [$value];
    }

}








